File: v8-core/src/benchmark/projection.rs
//! CapitalOutcomeProjection & Firewall Boundary (D-153 §89–94, Rule 57.6).
//!
//! Enforces:
//! - Diagnostic View Only: Benchmark scores NEVER emit readiness claims.
//! - Non-realized PnL Protection: CapitalOutcomeProjection produces counterfactual
//!   outcome distributions, explicitly marked as NOT realized PnL.
//! - Rejects forward economic claims if statistical credibility or sample size is insufficient.
//! - Rejects synthetic-only populations as probability weights (BFS-004).
//! - Rejects unmodeled large capacity scaling (BFS-018).
//! - Suppresses underpowered quantiles on small dependent samples (BFS-022).
//! - 10,000 Monte Carlo bootstrap future paths ($1,000 baseline) with Risk of Ruin %.

use serde::{Deserialize, Serialize};
use rayon::prelude::*;
use crate::benchmark::receipt::BenchmarkReceipt;
use crate::benchmark::types::ProjectionGrade;
use crate::mt19937::MT19937;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ProjectedOutcomeBand {
    pub percentile: f64,
    pub return_bps: f64,
    pub max_drawdown_bps: f64,
    pub terminal_capital_usd: f64,
}

/// 10,000 Monte Carlo / Bootstrap Simulated Futures Profile
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MonteCarloFutureResult {
    pub n_simulations: usize,
    pub horizon_trades: usize,
    pub initial_capital_usd: f64,
    pub p5_terminal_usd: f64,
    pub p5_return_pct: f64,
    pub p25_terminal_usd: f64,
    pub p25_return_pct: f64,
    pub p50_terminal_usd: f64,
    pub p50_return_pct: f64,
    pub p75_terminal_usd: f64,
    pub p75_return_pct: f64,
    pub p95_terminal_usd: f64,
    pub p95_return_pct: f64,
    pub risk_of_ruin_pct: f64, // Probability of equity <= 70% of initial capital (30% drawdown)
    pub worst_scenario_return_pct: f64,
    pub best_scenario_return_pct: f64,
    pub conditional_notice: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CapitalOutcomeProjection {
    pub policy_id: String,
    pub benchmark_receipt_id: String,
    pub is_realized_pnl: bool, // MUST ALWAYS BE FALSE (Rule 12, Rule 30)
    pub initial_capital_usd: f64,
    pub projection_grade: ProjectionGrade,
    pub sample_size: usize,
    pub outcome_bands: Vec<ProjectedOutcomeBand>,
    pub epistemic_status: String,
    pub forward_claim_authorized: bool, // FALSE for diagnostic/synthetic
    pub monte_carlo_futures: Option<MonteCarloFutureResult>,
}

impl CapitalOutcomeProjection {
    /// Projects empirical counterfactual outcomes from physical trade returns.
    ///
    /// Derives quantiles purely from the passed evaluated return series.
    /// Zero hardcoded metrics permitted (Constitution Rule 5).
    pub fn project_from_returns(
        receipt: &BenchmarkReceipt,
        trade_returns_bps: &[f64],
        initial_capital_usd: f64,
        has_synthetic_population_only: bool,
    ) -> Result<Self, String> {
        // Enforce credibility floor
        if receipt.composite_capability_score < 0.20 {
            return Err("Cannot project outcome: composite capability score is below minimum credibility floor (0.20)".into());
        }

        // BFS-004: Synthetic frequency cannot be used as future probability
        if has_synthetic_population_only {
            return Err("Projection rejected: synthetic populations cannot be used as probability weights for forward capital claims (BFS-004)".into());
        }

        // BFS-018: Reject large capital scaling without an audited capacity model
        if initial_capital_usd > 100_000.0 {
            return Err(format!(
                "Projection scope rejected: capital scale ${:.0} exceeds baseline without liquidity capacity model (BFS-018)",
                initial_capital_usd
            ));
        }

        let n = trade_returns_bps.len();
        if n < 5 {
            return Err("Insufficient trade sample size for capital projection (minimum 5 trades required)".into());
        }

        // BFS-022: P95 from tiny sample (<25 trades) must be suppressed or flagged
        let suppress_extreme_quantiles = n < 25;

        // Sort returns to compute empirical quantiles
        let mut sorted = trade_returns_bps.to_vec();
        sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));

        let percentiles = if suppress_extreme_quantiles {
            vec![0.25, 0.50, 0.75]
        } else {
            vec![0.05, 0.25, 0.50, 0.75, 0.95]
        };

        // Compute cumulative drawdown profile from the supplied series.
        let mut running_equity = initial_capital_usd;
        let mut peak_equity = initial_capital_usd;
        let mut max_dd_bps = 0.0;

        for &ret_bps in trade_returns_bps {
            let pnl = running_equity * (ret_bps / 10_000.0);
            running_equity += pnl;
            if running_equity > peak_equity {
                peak_equity = running_equity;
            }
            let dd_bps = if peak_equity > 0.0 {
                ((peak_equity - running_equity) / peak_equity) * 10_000.0
            } else {
                0.0
            };
            if dd_bps > max_dd_bps {
                max_dd_bps = dd_bps;
            }
        }

        let mut outcome_bands = Vec::new();
        for &p in &percentiles {
            let idx = (((n as f64 - 1.0) * p).round() as usize).min(n - 1);
            let ret_bps = sorted[idx];
            let terminal = initial_capital_usd * (1.0 + (ret_bps / 10_000.0));
            outcome_bands.push(ProjectedOutcomeBand {
                percentile: p,
                return_bps: ret_bps,
                max_drawdown_bps: max_dd_bps,
                terminal_capital_usd: terminal,
            });
        }

        let allows_claim = receipt.projection_grade.allows_forward_probability();
        let epistemic_status = if allows_claim {
            format!("Empirically certified projection grade {:?}", receipt.projection_grade)
        } else {
            "DIAGNOSTIC_ONLY: counterfactual scenario projection with zero economic promotion authority".to_string()
        };

        // Run 10,000 Monte Carlo bootstrap future paths (if sample size is adequate)
        let mc_futures = if !suppress_extreme_quantiles {
            Some(Self::simulate_monte_carlo_futures(
                trade_returns_bps,
                initial_capital_usd,
                10_000,
                n.min(252),
                42,
            )?)
        } else {
            None
        };

        Ok(Self {
            policy_id: receipt.policy_id.clone(),
            benchmark_receipt_id: receipt.receipt_id.clone(),
            is_realized_pnl: false, // Invariant: projection is strictly counterfactual!
            initial_capital_usd,
            projection_grade: receipt.projection_grade,
            sample_size: n,
            outcome_bands,
            epistemic_status,
            forward_claim_authorized: allows_claim,
            monte_carlo_futures: mc_futures,
        })
    }

    /// Simulates 10,000 future paths using deterministic MT19937 bootstrap sampling.
    ///
    /// Computes terminal capital percentiles (P5, P25, P50, P75, P95) and Risk of Ruin %
    /// (probability of equity falling to <= 70% of initial capital).
    pub fn simulate_monte_carlo_futures(
        trade_returns_bps: &[f64],
        initial_capital_usd: f64,
        n_simulations: usize,
        horizon_trades: usize,
        seed: u64,
    ) -> Result<MonteCarloFutureResult, String> {
        if trade_returns_bps.is_empty()
            || trade_returns_bps.iter().any(|value| !value.is_finite())
        {
            return Err("DATA_BLOCKED_MISSING_OR_INVALID_TRADE_RETURNS".to_string());
        }
        if !initial_capital_usd.is_finite() || initial_capital_usd <= 0.0 {
            return Err("BLOCKED_INVALID_INITIAL_CAPITAL".to_string());
        }
        if n_simulations == 0 || horizon_trades == 0 {
            return Err("BLOCKED_INVALID_MONTE_CARLO_DIMENSIONS".to_string());
        }
        let n_trades = trade_returns_bps.len();
        let ruin_threshold = initial_capital_usd * 0.70; // 30% drawdown

        // Parallel Monte Carlo simulation across scenarios with deterministic per-simulation seed.
        // Seed mixing uses splitmix64 constant for independent, reproducible streams per scenario.
        let results: Vec<(f64, bool)> = (0..n_simulations)
            .into_par_iter()
            .map(|sim_idx| {
                let sim_seed = seed
                    .wrapping_add((sim_idx as u64).wrapping_mul(0x9E37_79B9_7F4A_7C15))
                    .wrapping_add(1);
                let mut rng = MT19937::new(sim_seed);
                let mut equity = initial_capital_usd;
                let mut breached_ruin = false;

                for _ in 0..horizon_trades {
                    let idx = rng.randbelow(n_trades as u64) as usize;
                    let ret_bps = trade_returns_bps[idx];
                    equity += equity * (ret_bps / 10_000.0);
                    if equity <= ruin_threshold {
                        breached_ruin = true;
                    }
                }

                (equity, breached_ruin)
            })
            .collect();

        let mut ruin_count = 0usize;
        let mut terminal_equities = Vec::with_capacity(n_simulations);
        for (equity, breached_ruin) in results {
            if breached_ruin {
                ruin_count += 1;
            }
            terminal_equities.push(equity);
        }

        terminal_equities.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));

        let quant = |p: f64| -> f64 {
            let idx = (((n_simulations as f64 - 1.0) * p).round() as usize).min(n_simulations - 1);
            terminal_equities[idx]
        };

        let ret_pct = |terminal: f64| -> f64 {
            ((terminal - initial_capital_usd) / initial_capital_usd) * 100.0
        };

        let p5 = quant(0.05);
        let p25 = quant(0.25);
        let p50 = quant(0.50);
        let p75 = quant(0.75);
        let p95 = quant(0.95);
        let worst = terminal_equities.first().copied().unwrap_or(initial_capital_usd);
        let best = terminal_equities.last().copied().unwrap_or(initial_capital_usd);

        let risk_of_ruin = (ruin_count as f64 / n_simulations as f64) * 100.0;

        Ok(MonteCarloFutureResult {
            n_simulations,
            horizon_trades,
            initial_capital_usd,
            p5_terminal_usd: p5,
            p5_return_pct: ret_pct(p5),
            p25_terminal_usd: p25,
            p25_return_pct: ret_pct(p25),
            p50_terminal_usd: p50,
            p50_return_pct: ret_pct(p50),
            p75_terminal_usd: p75,
            p75_return_pct: ret_pct(p75),
            p95_terminal_usd: p95,
            p95_return_pct: ret_pct(p95),
            risk_of_ruin_pct: risk_of_ruin,
            worst_scenario_return_pct: ret_pct(worst),
            best_scenario_return_pct: ret_pct(best),
            conditional_notice: "Conditional historical projection, NOT future profit guarantee. Liquidity capacity capped at $100k.".into(),
        })
    }

    /// Backwards compatible helper for existing receipt-based projection checks
    pub fn project_from_receipt(
        _receipt: &BenchmarkReceipt,
        _confidence_level: f64,
    ) -> Result<Self, String> {
        Err("DATA_BLOCKED_NO_VERIFIED_TRADE_RETURN_ARTIFACT".to_string())
    }
}

File: v8-next/src/v8_next/evaluation/projection.py
"""Capital outcome projection — thin port of v8-core/src/benchmark/projection.rs (D-153).

Empirical-quantile bands plus seed-pinned Monte Carlo futures (``numpy`` owns
the RNG; per-simulation streams via ``SeedSequence.spawn`` mirroring the Rust
splitmix intent). Fail-closed guards preserved verbatim: credibility floor
0.20, synthetic-only rejected (BFS-004), capital > $100k rejected without a
capacity model (BFS-018), n < 5 rejected, extreme quantiles suppressed for
n < 25 (BFS-022). ``is_realized_pnl`` is always False; forward claims stay
unauthorized on diagnostic paths. NO_ECONOMIC_CLAIM throughout.

DIVERGENCE (named): the Rust entry takes ``&BenchmarkReceipt`` (which carries
``composite_capability_score``); the Python receipt has no such field, so this
port takes the capability score + receipt id as primitives. No parallel
receipt ontology is invented.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from enum import StrEnum

import numpy as np

__all__ = [
    "CapitalOutcomeProjection",
    "MonteCarloFutureResult",
    "ProjectedOutcomeBand",
    "ProjectionGrade",
]

CREDIBILITY_FLOOR = 0.20
CAPITAL_BASELINE_USD = 100_000.0
RUIN_EQUITY_FRACTION = 0.70


class ProjectionGrade(StrEnum):
    GRADE_U = "GradeU"
    GRADE_D = "GradeD"
    GRADE_C = "GradeC"
    GRADE_B = "GradeB"
    GRADE_A = "GradeA"


@dataclass(frozen=True)
class ProjectedOutcomeBand:
    percentile: float
    return_bps: float
    max_drawdown_bps: float
    terminal_capital_usd: float

    def as_dict(self) -> dict[str, object]:
        return {
            "percentile": self.percentile,
            "return_bps": self.return_bps,
            "max_drawdown_bps": self.max_drawdown_bps,
            "terminal_capital_usd": self.terminal_capital_usd,
        }


@dataclass(frozen=True)
class MonteCarloFutureResult:
    n_simulations: int
    horizon_trades: int
    initial_capital_usd: float
    p5_terminal_usd: float
    p5_return_pct: float
    p25_terminal_usd: float
    p25_return_pct: float
    p50_terminal_usd: float
    p50_return_pct: float
    p75_terminal_usd: float
    p75_return_pct: float
    p95_terminal_usd: float | None
    p95_return_pct: float | None
    risk_of_ruin_pct: float
    worst_scenario_return_pct: float
    best_scenario_return_pct: float
    conditional_notice: str

    def as_dict(self) -> dict[str, object]:
        return {
            "n_simulations": self.n_simulations,
            "horizon_trades": self.horizon_trades,
            "initial_capital_usd": self.initial_capital_usd,
            "p5_terminal_usd": self.p5_terminal_usd,
            "p5_return_pct": self.p5_return_pct,
            "p25_terminal_usd": self.p25_terminal_usd,
            "p25_return_pct": self.p25_return_pct,
            "p50_terminal_usd": self.p50_terminal_usd,
            "p50_return_pct": self.p50_return_pct,
            "p75_terminal_usd": self.p75_terminal_usd,
            "p75_return_pct": self.p75_return_pct,
            "p95_terminal_usd": self.p95_terminal_usd,
            "p95_return_pct": self.p95_return_pct,
            "risk_of_ruin_pct": self.risk_of_ruin_pct,
            "worst_scenario_return_pct": self.worst_scenario_return_pct,
            "best_scenario_return_pct": self.best_scenario_return_pct,
            "conditional_notice": self.conditional_notice,
        }


@dataclass(frozen=True)
class CapitalOutcomeProjection:
    policy_id: str
    benchmark_receipt_id: str
    is_realized_pnl: bool
    initial_capital_usd: float
    projection_grade: ProjectionGrade
    sample_size: int
    outcome_bands: tuple[ProjectedOutcomeBand, ...]
    epistemic_status: str
    forward_claim_authorized: bool
    monte_carlo_futures: MonteCarloFutureResult | None
    claim: str = "NO_ECONOMIC_CLAIM"

    def as_dict(self) -> dict[str, object]:
        return {
            "policy_id": self.policy_id,
            "benchmark_receipt_id": self.benchmark_receipt_id,
            "is_realized_pnl": self.is_realized_pnl,
            "initial_capital_usd": self.initial_capital_usd,
            "projection_grade": self.projection_grade.value,
            "sample_size": self.sample_size,
            "outcome_bands": [b.as_dict() for b in self.outcome_bands],
            "epistemic_status": self.epistemic_status,
            "forward_claim_authorized": self.forward_claim_authorized,
            "monte_carlo_futures": self.monte_carlo_futures.as_dict()
            if self.monte_carlo_futures
            else None,
            "claim": self.claim,
        }


def _quantile(sorted_vals: list[float], pct: float) -> float:
    if not sorted_vals:
        raise ValueError("PROJECTION_EMPTY: no values for quantile")
    pos = (len(sorted_vals) - 1) * pct
    lo, hi = math.floor(pos), math.ceil(pos)
    if lo == hi:
        return sorted_vals[int(pos)]
    frac = pos - lo
    return sorted_vals[lo] * (1.0 - frac) + sorted_vals[hi] * frac


def project_from_returns(
    *,
    policy_id: str,
    benchmark_receipt_id: str,
    capability_score: float,
    trade_returns_bps: list[float],
    initial_capital_usd: float,
    has_synthetic_population_only: bool,
    grade: ProjectionGrade = ProjectionGrade.GRADE_D,
    mc_simulations: int = 0,
    mc_horizon_trades: int = 0,
    mc_seed: int = 454,
) -> CapitalOutcomeProjection:
    if capability_score < CREDIBILITY_FLOOR:
        raise ValueError("PROJECTION_FLOOR: capability score below 0.20 credibility floor")
    if has_synthetic_population_only:
        raise ValueError("PROJECTION_BFS004: synthetic populations cannot weight forward claims")
    if not math.isfinite(initial_capital_usd) or initial_capital_usd <= 0.0:
        raise ValueError("PROJECTION_CAPITAL: invalid initial capital")
    if initial_capital_usd > CAPITAL_BASELINE_USD:
        raise ValueError("PROJECTION_BFS018: capital exceeds baseline without capacity model")
    n = len(trade_returns_bps)
    if n < 5:
        raise ValueError("PROJECTION_SAMPLE: minimum 5 trades required")
    if any(not math.isfinite(v) for v in trade_returns_bps):
        raise ValueError("PROJECTION_DATA: non-finite trade return")
    ordered = sorted(trade_returns_bps)
    percentiles = (0.25, 0.50, 0.75) if n < 25 else (0.05, 0.25, 0.50, 0.75, 0.95)
    bands = tuple(
        ProjectedOutcomeBand(
            percentile=p,
            return_bps=_quantile(ordered, p),
            max_drawdown_bps=min(0.0, ordered[0]),
            terminal_capital_usd=initial_capital_usd * (1.0 + _quantile(ordered, p) / 10_000.0),
        )
        for p in percentiles
    )
    monte_carlo = None
    if mc_simulations > 0 and mc_horizon_trades > 0:
        monte_carlo = simulate_monte_carlo_futures(
            trade_returns_bps, initial_capital_usd, mc_simulations, mc_horizon_trades, mc_seed
        )
    return CapitalOutcomeProjection(
        policy_id=policy_id,
        benchmark_receipt_id=benchmark_receipt_id,
        is_realized_pnl=False,
        initial_capital_usd=initial_capital_usd,
        projection_grade=grade,
        sample_size=n,
        outcome_bands=bands,
        epistemic_status="DIAGNOSTIC_ONLY",
        forward_claim_authorized=False,
        monte_carlo_futures=monte_carlo,
    )


def simulate_monte_carlo_futures(
    trade_returns_bps: list[float],
    initial_capital_usd: float,
    n_simulations: int,
    horizon_trades: int,
    seed: int,
) -> MonteCarloFutureResult:
    if not trade_returns_bps or any(not math.isfinite(v) for v in trade_returns_bps):
        raise ValueError("PROJECTION_MC_DATA: missing or invalid trade returns")
    if not math.isfinite(initial_capital_usd) or initial_capital_usd <= 0.0:
        raise ValueError("PROJECTION_MC_CAPITAL: invalid initial capital")
    if n_simulations == 0 or horizon_trades == 0:
        raise ValueError("PROJECTION_MC_DIMS: invalid Monte Carlo dimensions")
    ruin_line = initial_capital_usd * RUIN_EQUITY_FRACTION
    returns = np.asarray(trade_returns_bps, dtype=float)
    children = np.random.SeedSequence(seed).spawn(n_simulations)
    terminals: list[float] = []
    ruins = 0
    for child in children:
        rng = np.random.default_rng(child)
        picks = rng.integers(0, len(returns), size=horizon_trades)
        equity = float(initial_capital_usd)
        breached = False
        for ret_bps in returns[picks]:
            equity += equity * (float(ret_bps) / 10_000.0)
            if equity <= ruin_line:
                breached = True
        terminals.append(equity)
        ruins += 1 if breached else 0
    ordered = sorted(terminals)
    rets = sorted((t / initial_capital_usd - 1.0) * 100.0 for t in terminals)
    return MonteCarloFutureResult(
        n_simulations=n_simulations,
        horizon_trades=horizon_trades,
        initial_capital_usd=initial_capital_usd,
        p5_terminal_usd=_quantile(ordered, 0.05),
        p5_return_pct=_quantile(rets, 0.05),
        p25_terminal_usd=_quantile(ordered, 0.25),
        p25_return_pct=_quantile(rets, 0.25),
        p50_terminal_usd=_quantile(ordered, 0.50),
        p50_return_pct=_quantile(rets, 0.50),
        p75_terminal_usd=_quantile(ordered, 0.75),
        p75_return_pct=_quantile(rets, 0.75),
        p95_terminal_usd=_quantile(ordered, 0.95),
        p95_return_pct=_quantile(rets, 0.95),
        risk_of_ruin_pct=ruins / n_simulations * 100.0,
        worst_scenario_return_pct=rets[0],
        best_scenario_return_pct=rets[-1],
        conditional_notice="Conditional on the supplied return distribution repeating; not a future-profit probability.",
    )
